diagnostics-extension/src/controllers/diagnostics.ts

```diff
@@ -6,14 +6,112 @@
+} from '../utils';
 
 type DiagnosticsController = (
   category: Request,
   res: (data: Response) => void
 ) => void;
 
+const diagnosticsService = DiagnosticsServiceProvider.getDiagnosticsService();
+
+const handleStartRoutine = async (
+  routineName: DiagnosticsRoutineName | undefined,
+  res: (data: Response) => void,
+  params?: DiagnosticsParams
+) => {
+  if (!routineName) {
+    return res(
+      generateErrorResponse(
+        ResponseErrorInfoMessage.InvalidDiagnosticsRoutineName
+      )
+    );
+  }
+  const routineId = diagnosticsService.runRoutine(routineName, params);
+  const routineStatus = await diagnosticsService.getRoutineStatus(routineId);
+  const payload: DiagnosticsResponse = { routineId, routineStatus };
+  return res(generateDiagnosticsSuccessResponse(payload));
+};
+
+const handleStopRoutine = async (
+  routineId: number | undefined,
+  res: (data: Response) => void
+) => {
+  if (!routineId) {
+    return res(
+      generateErrorResponse(
+        ResponseErrorInfoMessage.InvalidDiagnosticsRoutineId
+      )
+    );
+  }
+  const routineStatus = await diagnosticsService.stopRoutine(routineId);
+  const payload: DiagnosticsResponse = { routineId, routineStatus };
+  return res(generateDiagnosticsSuccessResponse(payload));
+};
+
+const handleGetRoutineStatus = async (
+  routineId: number | undefined,
+  res: (data: Response) => void
+) => {
+  if (!routineId) {
+    return res(
+      generateErrorResponse(
+        ResponseErrorInfoMessage.InvalidDiagnosticsRoutineId
+      )
+    );
+  }
+  const routineStatus = await diagnosticsService.getRoutineStatus(routineId);
+  const payload: DiagnosticsResponse = { routineId, routineStatus };
+  return res(generateDiagnosticsSuccessResponse(payload));
+};
+
+const handleResumeRoutine = async (
+  routineId: number | undefined,
+  res: (data: Response) => void
+) => {
+  if (!routineId) {
+    return res(
+      generateErrorResponse(
+        ResponseErrorInfoMessage.InvalidDiagnosticsRoutineId
+      )
+    );
+  }
+  const routineStatus = await diagnosticsService.resumeRoutine(routineId);
+  const payload: DiagnosticsResponse = { routineId, routineStatus };
+  return res(generateDiagnosticsSuccessResponse(payload));
+};
+
 export const handleDiagnostics: DiagnosticsController = (req, res) => {
-  res(generateErrorResponse('Diagnostics not implemented.'));
+  if (!req.diagnostics)
+    return res(
+      generateErrorResponse(ResponseErrorInfoMessage.MissingDiagnosticsRequest)
+    );
+
+  switch (req.diagnostics.action) {
+    case DiagnosticsAction.START:
+      return handleStartRoutine(req.diagnostics.routineName, res, req.diagnostics.params);
+    case DiagnosticsAction.STATUS:
+      return handleGetRoutineStatus(req.diagnostics.routineId, res);
+    case DiagnosticsAction.RESUME:
+      return handleResumeRoutine(req.diagnostics.routineId, res);
+    case DiagnosticsAction.STOP:
+      return handleStopRoutine(req.diagnostics.routineId, res);
```

Comment:
The only difference between the 3 methods (handleGetRoutineStatus, handleResumeRoutine, handleStopRoutine) is the call to the diagnosticsService.{getStatus, resume, stop}. You can directly map the enum value to the method to be called:

```
const routineOperationMap: ... = {
  DiagnosticsAction.STATUS: diagnosticsService.getRoutineStatus,
  DiagnosticsAction.RESUME: diagnosticsService.resumeRoutine,
  DiagnosticsAction.STOP: diagnosticsService.stopRoutine,
}
..
..
// this switch
switch (req.diagnostics.action) {
    case DiagnosticsAction.START:
      return handleStartRoutine(req.diagnostics.routineName, res, req.diagnostics.params);
    case DiagnosticsAction.STATUS:
    case DiagnosticsAction.RESUME:
    case DiagnosticsAction.STOP:
      return handleRoutineOperation(req.diagnostics.routineId, routineOperationMap[req.diagnostics.action]);
  default:
      return res(
        generateErrorResponse(ResponseErrorInfoMessage.InvalidDiagnosticsAction)
      );
  }
```
and then, you end up implementing one handleRoutineOperation() method and not 3:
```
const handleRoutineOperation = async (
  routineId: number | undefined,
  routineOperation: Promise<RoutineStatus>,
  res: (data: Response) => void
) => {
  if (!routineId) {
    return res(
      generateErrorResponse(
        ResponseErrorInfoMessage.InvalidDiagnosticsRoutineId
      )
    );
  }
  const routineStatus = await routineOperation(routineId);
  const payload: DiagnosticsResponse = { routineId, routineStatus };
  return res(generateDiagnosticsSuccessResponse(payload));
};
```

Moreover, you can reduce handleStartRoutine() as follows:
```
const handleStartRoutine = async (
  routineName: DiagnosticsRoutineName | undefined,
  res: (data: Response) => void,
  params?: DiagnosticsParams
) => {
  if (!routineName) {
    return res(
      generateErrorResponse(
        ResponseErrorInfoMessage.InvalidDiagnosticsRoutineName
      )
    );
  }
  const routineId = diagnosticsService.runRoutine(routineName, params);
  return handleRoutineOperation(routineId, diagnosticsService.getRoutineStatus, res);
};
```